Disallow Password Change when authenticated by Token

jkakavas opened the pull request:
This commit explicitly disallows Change Password requests
when the user performing the request is authenticated with
an Elasticsearch TokenService access token or an API key.

We should coordinate merging this with kibana since the password change feature would stop working when the token authentication provider is in use in Kibana. 

Relates: https://github.com/elastic/kibana/issues/49865
Closes: https://github.com/elastic/elasticsearch/issues/48752

jkakavas committed:
Mark token authN in the authentication object

This commit changes that behavior so that we set the authenticatedBy
realm reference to _es_token_service in the Authentication object
we build and write to the thread context upon successful
authentication with an Elasticsearch Token Service access token.
This alligns our behavior with the API keys implementation and
allows us to make authorization decisions based on the fact that
the current request was authenticated by a bearer token ( i.e.
disallow change password requests ).

The original authentication realm is still available in the token
document.

elasticmachine commented:
Pinging @elastic/es-security (:Security/Authentication)

jkakavas committed:
Merge remote-tracking branch 'origin/master' into disallow-pwd-change-with-token

tvernum commented:
Have we considered any other BWC issues with changing this?
I don't specifically know of any, but it's in that grey area where I can imagine there might be some.

We could resolve the password change behaviour by looking at the `AuthenticationType` as well as the realm (which would avoid needing this change).

jkakavas commented:
> We could resolve the password change behaviour by looking at the AuthenticationType as well as the realm (which would avoid needing this change).

True, this is not the only way to solve the password change behavior but I like it since it's consistent with our behavior for API keys. I looked for other uses of `authenticatedBy` and didn't find any potential issues. I can go through them and list them in a way that it makes it easier for us to argue about potential BWC issues (or the lack thereof).

jkakavas commented:
Let's discuss this on Thursday morning, I want to get a team view on how we treat/should treat the authenticatedBy attribute of the Authentication object when tokens are used by our own realms ( DelegeatedPKI, OIDC, SAML)

jkakavas committed:
Fix test

jkakavas committed:
Disallow based on AuthenticationType

Changes the approach for disallowing password changes to check the
AuthenticationType instead of injecting a custom RealmRef for tokens
as the authenticatedBy Realm Reference

jkakavas commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into disallow-pwd-change-with-token

jkakavas committed:
revert test change

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:
```diff
@@ -524,9 +524,13 @@ private static boolean checkChangePasswordAction(Authentication authentication)
+        // and right now only one can exist in the realm configuration - if this changes we should update this check
+        final Authentication.AuthenticationType authType = authentication.getAuthenticationType();
+        return (authType.equals(Authentication.AuthenticationType.TOKEN) == false
+        && authType.equals(Authentication.AuthenticationType.API_KEY) == false )
```
Why do we blacklist rather than whitelist here?
Shouldn't we just check for the positive `authType.equals(Authentication.AuthenticationType.REALM)` ?

jkakavas replied:
You're absolutely right

jkakavas committed:
Merge remote-tracking branch 'origin/master' into disallow-pwd-change-with-token

jkakavas committed:
whitelist>blacklist

tvernum reviewed (approved):
LGTM.
I haven't tracked the status of the Kibana side of the change, so I'll leave it to you to work out whether it should be in 7.6 or wait for 7.7

polyfractal commented:
@jkakavas Is this still `backport_pending` for 7.6, or bumped to 7.7?
